Combine Python types in PackedInstructions

mtreinish opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit migrates the 3 python variants in PackedOperation to be a single variant. The 3 types PyGate, PyInstruction, and PyOperation are very similar and are only separate for typing reasons. This commit moves to only having a single struct and adding the typing information as an enum that wraps that struct representation. This is then propogated to the PackedInstruction type and move us from 3 separate fields in the struct to just a single one. This frees up more space for other operation types in the future. OperationRef keeps the distinct variants for Gate, Instruction, and Operation. This simplifies the required changes as most places interact with OperationRef instead of the owned types.

The other motivation for this is as an early step towards fixing #14240. This commit simplifies and consolidates the places that explicitly wrap python usage in the circuit crate. This will make it easier to conditionally compile things when we start to do that.

The other option considered was also moving to a single struct and setting a `py_op_type` field inside the struct that was an enum to differentiate the Python type. However, doing this inside the struct would require significant changes to the usage of Python types in a circuit. Instead of matching on the type off of OperationRef it would need to be a double mapping first on OperationRef then on `py_op_type` field. Using the outer enum prevents this and makes it a bit easier to match in places.

### Details and comments

mtreinish committed:
Combine Python types in PackedInstructions

This commit migrates the 3 python variants in PackedOperation to be a
single variant. The 3 types PyGate, PyInstruction, and PyOperation are
very similar and are only separate for typing reasons. This commit moves
to only having a single struct and adding the typing information as an
enum that wraps that struct representation. This is then propogated to
the PackedInstruction type and move us from 3 separate fields in the
struct to just a single one. This frees up more space for other
operation types in the future. OperationRef keeps the distinct variants
for Gate, Instruction, and Operation. This simplifies the required
changes as most places interact with OperationRef instead of the owned
types.

The other motivation for this is as an early step towards fixing #14240.
This commit simplifies and consolidates the places that explicitly wrap
python usage in the circuit crate. This will make it easier to
conditionally compile things when we start to do that.

The other option considered was also moving to a single struct and setting
a `py_op_type` field inside the struct that was an enum to differentiate
the Python type. However, doing this inside the struct would require
significant changes to the usage of Python types in a circuit. Instead of
matching on the type off of OperationRef it would need to be a double
mapping first on OperationRef then on `py_op_type` field. Using the
outer enum prevents this and makes it a bit easier to match in places.

raynelfss reviewed (dismissed):
This looks good to me, I think it makes sense to combine all three python gate types into one struct, and we don't seem to be losing any information in the process.

kevinhartman commented:
One thing is, it might be nice if control flow merges before this one? @jakelishman

Shobhit21287 commented on crates/circuit/src/operations.rs:
```diff
@@ -2455,6 +2455,13 @@ pub trait PythonOperation: Sized {
 }
 
+#[derive(Clone, Debug)]
+pub enum PyOperationTypes {
```
Hi Matt, Its really nice what you've done to handle all the python types as a single enum. I just had one question, the macro  `#[repr(align(8))]` which I assume is the macro for bit alignment is not declared for `PyOperationTypes` but rather `PyInstruction`. The `PackedOperation` however, looks as follows.
```rust 
enum PackedOperationType {
    StandardGate = 0,
    StandardInstruction = 1,
    PyOperationTypes = 2,
    UnitaryGate = 3,
}
```
I was wondering if the macro should be declared over `PyOperationTypes`?  Wasn’t sure if I was missing something about how the packing or memory layout works for this.

mtreinish replied:
Sorry for the slow response here I missed this review at first. The reason we don't need it here is because the alignment is needed for the pointer type to support how `PackedOperation` works. The `align(8)` enables us to pack a discriminant in the unused bits in the pointer (see the docs for packed operation for more details on this), but for this type we don't need to specify the alignment because it's not used in the packed operation. Since `PyInstruction` is part of the packed operation it needs to have `align(8)` set, but for this type enum it doesn't matter what the alignment is because it's just used as a type discriminant inside `PyInstruction` and we can let the compiler pick whatever alignment it thinks is best.

mtreinish replied:
Actually you're correct I misremembered the PR because it had been a while. We want to force an alignment on `PyOperationTypes` to since that is what we put in the packedoperation. I was remembering an earlier iteration of this PR before I pushed it that used it as a struct field in PyInstruction before moving to this model. I'll fix this now.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/15277/commits/4377c977e04c77e3e2e696d7d8c8cb24fb3c16a5

mtreinish committed:
Merge remote-tracking branch 'origin/main' into simplify-py-op-structs

mtreinish committed:
Add missing alignment on PyOperationTypes

As we insert PyOperationTypes in the PackedInstruction we need to ensure
the alignment so that there are free bits to put the discriminant of the
PackedOperation in the u64. This was overlooked in earlier versions of
this PR branch and would have caused issues on other platforms
potentially.

mtreinish committed:
Fix typo in rebase

mtreinish committed:
Merge remote-tracking branch 'origin/main' into simplify-py-op-structs

Cryoris reviewed (approved):
I think this is a good step to encapsulate our Python dependency more (and as small convenience also frees up space in `PackedOperation` we can use), but there's the tradeoff that we can have former `PyGate`-specific methods now available on all Python types (-> `matrix`, `num_ctrl_qubits`, ...). 

Given that this was the state before those methods were removed from the `Operation` trait, and the Python-backed objects need care anyways I'm fine moving ahead with this though. 

Since Ray already approved before I'll put it in the merge queue right away
